fbw-a380x/src/wasm/systems/a380_systems/src/fuel/fuel_quantity_management_system.rs

```diff
@@ -237,10 +307,97 @@ impl RefuelApplication {
         }
     }
 
+    fn lookup_trim_fuel_from_target_fuel_range(
+        &mut self,
+        category_index: usize,
+        total_fuel_desired_rounded: u32,
+        zero_fuel_weight_cg_rounded: u32,
+    ) -> Mass {
+        let zfw_category = &self.trim_tank_map.trim_tank_targets.tables[category_index];
+        let target_fuel_range = zfw_category.get(&total_fuel_desired_rounded).unwrap();
+
+        if let Some(cg_index) = self
+            .trim_tank_map
+            .params
+            .target_zfw_cg
+            .iter()
+            .position(|&x| x == zero_fuel_weight_cg_rounded)
+        {
+            return Mass::new::<kilogram>(target_fuel_range[cg_index] as f64);
+        } else {
+            return Mass::new::<kilogram>(target_fuel_range[0] as f64);
+        }
+    }
+
+    fn lookup_trim_fuel_from_mapping(
+        &mut self,
+        category_index: usize,
+        total_desired_fuel_rounded: u32,
+        zero_fuel_weight_cg_rounded: u32,
+    ) -> Mass {
+        let zfw_category = &self.trim_tank_map.trim_tank_targets.tables[category_index];
+        let lowest_fuel_desired = *(zfw_category.keys().min().unwrap());
+        let largest_fuel_desired = *(zfw_category.keys().max().unwrap());
+
+        match total_desired_fuel_rounded {
+            x if x < lowest_fuel_desired => self.lookup_trim_fuel_from_target_fuel_range(
+                category_index,
+                lowest_fuel_desired,
+                zero_fuel_weight_cg_rounded,
+            ),
+            x if x > largest_fuel_desired => self.lookup_trim_fuel_from_target_fuel_range(
+                category_index,
+                largest_fuel_desired,
+                zero_fuel_weight_cg_rounded,
+            ),
+            _ => self.lookup_trim_fuel_from_target_fuel_range(
+                category_index,
+                total_desired_fuel_rounded,
+                zero_fuel_weight_cg_rounded,
+            ),
+        }
+    }
+
+    fn calculate_trim_fuel(
+        &mut self,
+        total_desired_fuel: Mass,
+        zero_fuel_weight: Mass,
+        zero_fuel_weight_cg_percent_mac: f64,
+    ) -> Mass {
+        let total_desired_fuel_rounded =
+            ((total_desired_fuel.get::<kilogram>() / 1000.0).floor() * 1000.0) as u32;
+        let zero_fuel_weight_cg_rounded = (zero_fuel_weight_cg_percent_mac).floor() as u32;
+
+        let mut trim_fuel = Mass::default();
+
+        for (index, range) in self.trim_tank_map.params.target_zfw.iter().enumerate() {
+            if zero_fuel_weight >= Mass::new::<kilogram>(range.start as f64)
+                && zero_fuel_weight <= Mass::new::<kilogram>(range.end as f64)
+            {
+                trim_fuel = self.lookup_trim_fuel_from_mapping(
+                    index,
+                    total_desired_fuel_rounded,
+                    zero_fuel_weight_cg_rounded,
+                );
+                break;
+            }
+        }
+
+        trim_fuel
```

Comment:
```suggestion
        for (range, zfw_category) in self
            .trim_tank_map
            .params
            .target_zfw
            .iter()
            .zip(&self.trim_tank_map.trim_tank_targets.tables)
        {
            if zero_fuel_weight >= Mass::new::<kilogram>(range.start as f64)
                && zero_fuel_weight <= Mass::new::<kilogram>(range.end as f64)
            {
                return self.lookup_trim_fuel_from_mapping(
                    zfw_category,
                    total_desired_fuel_rounded,
                    zero_fuel_weight_cg_rounded,
                );
            }
        }

        Mass::default()
```
You'll have to adjust the functions `lookup_trim_fuel_from_mapping` and `lookup_trim_fuel_from_target_fuel_range` accordingly.

**Reasoning:** This way no out of bound access can happen.

Reply:
Too many borrows according to rust linting while still in the loop, have to kick out some values and clone them.